Fix sqlite3 fetchArray #64531

frikiluser opened the pull request:
Fix  #64531 SQLite3 SQLite3Result::fetchArray() 
Avoid sqlite3_step() call when result contains no columns (ie non-select-like query).

sqlite3_column_count() described here https://sqlite.org/c3ref/column_count.html
`A SELECT statement will always have a positive sqlite3_column_count()`

Patch uploaded: https://bugs.php.net/patch-display.php?bug_id=64531&patch=fix64531-sqlite3-fetchArray-skipNoColumns&revision=1580516014

cmb69 commented on ext/sqlite3/sqlite3.c:
```diff
@@ -2028,6 +2028,10 @@ PHP_METHOD(sqlite3result, fetchArray)
 	}
 
+	if (sqlite3_column_count(result_obj->stmt_obj->stmt) == 0) {
+		return;
```
I think we should `RETURN_FALSE` here, and also set `result_obj->complete = 1`. What do you think?

frikiluser replied:
Can you take a look on that commit?
 https://github.com/frikiluser/php-src/commit/a7529d63026b3cca7c9ce3d6acff00bdd4573f54

Changes from prev version: on top of php-7.3 branch, RETURN_FALSE, and completely whipped out result->complete handling.

I'll try to write test cases as requested.

cmb69 replied:
Well, the existing `->complete` handling doesn't look quite right (apparently, the `SQLITE_DONE` cases in `::query()` and `::execute()` don't set it), but replacing it by checking `sqlite3_column_count()` for `0` seems to change its semantics; checking for `sqlite3_data_count()` instead might be a good solution, but that function is only available as of sqlite3 3.6.10/3.7.2, but PHP 7.3 is supposed to be compatible with sqlite3 >= 3.3.9. So at least for PHP 7.3 we have to stick with `->complete`, I think. Also we cannot remove the `complete` member for ABI BC reasons in PHP 7.4, but we just could stop using it there. Anyhow, it seems to me that this is a different issue from bug #64531, so likely deserves its own ticket in the bugtracker.

frikiluser replied:
I've just pushed a commit with requested changes.
Test included showing what is fixed.

Here: https://github.com/frikiluser/php-src/commit/e8f8af30ad94442cbdd88d715b006c2809aa3eb5

Thanks for reviewing!

cmb69 replied:
Thanks for following up on this! It seems, though, that you forgot to push the test case. Could you please have a look?

frikiluser replied:
Hi.

I think it's already on my commit: https://github.com/frikiluser/php-src/commit/e8f8af30ad94442cbdd88d715b006c2809aa3eb5

New file: ext/sqlite3/tests/sqlite3result_fetchArray_nonrows_problem.phpt

Now I've updated the PR branch. Sorry for the mess up :-p

cmb69 replied:
> […] checking for sqlite3_data_count() instead might be a good solution, but that function is only available as of sqlite3 3.6.10/3.7.2 […]

Correction: sqlite3_data_count() is available at least as of 3.3.9 (likely even longer).

frikiluser commented:
Pushed again to force travis-ci tests rerun. Network problems...

frikiluser commented:
> Thanks for the PR! According to your [comment in the bugtracker](https://bugs.php.net/bug.php?id=64531#1580513024), this should probably target PHP-7.3, so please rebase onto that branch.

Note I've rebased from 7.3 branch and renamed as master in my repo. The pull request still points to php/master branch (should be changed to 7.3).

Let me know if I should open a new PR to the correct branch.

Regards

cmb69 commented:
> Let me know if I should open a new PR to the correct branch.

Nope, not necessary. I'll currently investigate on some not directly related issues I've stumbled upon while having a closer look at this PR. Will come back to this PR afterwards. Thanks for the added test case, anyway. :)

cmb69 commented:
I'm not sure if this solution would really be helpful. While it certainly catches some cases, it wouldn't work for the `ALTER TABLE` query mentioned in [bug #78813](https://bugs.php.net/bug.php?id=78813), and I'm not even sure if `sqlite3_column_count()` may be called after `sqlite3_reset()`, but that would happen for the first `::fetchArray()`.

PR #5204 would be a cleaner solution, but has issues on its own (error code).
